### PR TITLE
Fix width inferencing issue

### PR DESCRIPTION
**Pull request by jackkoenig**
As part of the Bundle literals framework PR, (https://github.com/freechipsproject/chisel3/pull/820) it appears a minor code refactor accidentally changed the width inference semantics of `WireInit` and `RegInit`.

For example
```scala
val wire1 = Wire(UInt(8.W))
val wire2 = WireInit(wire1)
```

prior to #820, wire2 would have it's width inferred. After #820, wire2 gets its width set to 8.

This PR restores the previous behavior and adds some tests that pass prior to #820, fail on current master, and pass on this branch

I noticed this issue while trying to bump Chisel because it changed the functionality of the output FIRRTL ever so slightly but enough to make simulations fail.

Future work is to make similar tests that show and check the width inferencing semantics for other constructors (eg. `RegNext`, `RegEnable`), as well as for aggregates (rather than just `UInt`).

**Related issue**: 

**Type of change**: bug report & fix

**Impact**: Undo API modification

**Development Phase**: implementation

**Release Notes**
Restore use of width inference to set width of RegInit and WireInit when initialized with a non-literal Bits

### CODE REVIEW

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 1]
```diff
@@ -507,8 +507,8 @@ object Wire extends WireFactory
 object WireInit {
   def apply[T <: Data](init: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
     val model = (init match {
-      // For e.g. Wire(init=0.U(k.W)), fix the Reg's width to k
```
Comment (aswaterman) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Thanks for updating the comment.  One reason this was able to regress is that I only wrote the comment to cover the special case, and didn't explain the intended semantics for the other cases.

**Comment by ducky64** [comment]
Can you write ScalaDoc that has a summary of the width inference rules for each function?

Off the top of my head I don't see strong reasons for one being more intuitive than the other, so it makes sense to keep compatible behavior, but I also think that being able to express the logic succinctly and intuitively is a good sniff test and benchmark.

**Comment by ucbjrl** [comment]
This PR (and anything after freechipsproject/firrtl#952) will fail Jenkins build due to the new methods required for the FIRRTL `Expression` abstract class which currently aren't implemented in the firrtl-interpreter which is built as part of Jenkins chisel3 builds.

**Comment by jackkoenig** [comment]
@ducky64 ScalaDoc and more tests (namely for `Reg` and `Wire`) added

**Comment by jackkoenig** [comment]
I'm open to better wording but AFAIK there really are the 3 cases. Literal Bits (set if width forced), Bits (inferred), and Aggregates (cloned from the Aggregate).

I clarified the ScalaDoc a little bit, and added some to the actual `apply` methods, including the neat `@usecase` tag which lets us hide the implicits! Assuming all is okay I'm going to merge this afternoon

**Comment by ducky64** [comment]
Yeah, I think it's nicer to have a very concise version, which you can always elaborate (eg edge cases) with illustrative examples. I'm more thinking of what gets a fixed width (only literals with defined width) or not (everything else it seems?).

Should something on bundle literals also be included in the inference rules?

**Comment by jackkoenig** [comment]
I added a shorter summary for single-argument WireInit and RegInit.

We should probably talk about bundle literals, but I'm thinking that can be follow on. I need this bug fix to bump chisel3 in rocket-chip so I would like to merge as soon as possible

**Pull request merged**